### ISSUE TITLE
stack 0.1.8 lost ability to build individual executables or library

### ISSUE THREAD

**Issue by ghorn** [commercialhaskell/stack#1406]
I have a cabal project with a library and many executables. In stack 0.1.6, I could 

```
stack build myproject:lib
```

to only build the library, or

```
stack build myproject:myexe1
stack build myproject:myexe2
```

to build individual executables. When I upgraded to stack 0.1.8 I lost this feature.

Losing this ability makes it take considerably longer to develop with stack, since I usually want to repeatedly rebuild one executable until it works, and having stack build all of them takes a very long time.

```
$ stack --version
Version 0.1.8.0, Git revision e33dd995e292efcf8ee42e1e4c649b96232abf2e (2603 commits) x86_64
```

**Comment by mgsloan** [comment]
Yeah, this was an unfortunate consequence of switching from running `install` to running `copy` + `register` (see #1203).  [My fix](https://github.com/commercialhaskell/stack/commit/dc6313f16bf9eb09f0607718cef1dab608b5705c) for this was to always build the executables with the library..

I agree this is rather unfortunate.  I'm not sure what to do about it, though.  Looks like we have a tradeoff between working around a Cabal issue in order to run `Setup.hs` the same way as cabal install, or support building individual executables.  You can try out different alternatives by running `stack build -v` and copy + pasting the lines that start with `Run process:`.

**Comment by mgsloan** [comment]
I've put a fix on this on master.  I consider this bug to be pretty substantial, so I'm in favor of doing a bugfix release.

**Comment by massysett** [comment]
Also, the `:lib` target is not documented in `doc/build_command.md`.

**Comment by mgsloan** [comment]
Bumping priority, since others are reporting it and it breaks our tutorials https://github.com/commercialhaskell/stack/issues/1554 .

Plus, the decision for #1203 seems to be somewhat arbitrary.  I think that stack *was* running `postInstall` before (via `runhaskell Setup.hs install`), and this is an arbitrary switch to using `postCopy`.  This needs to be investigated further, though.  Things also get tricky because Cabal-1.18 has different postInstall / postCopy behavior...

**Comment by snoyberg** [comment]
Just weighing in, since I saw this referenced from Reddit. The Cabal bug of erroring out when executables weren't built during copy/install exists regardless of whether we use `Setup install` or `Setup copy`, so the change referenced above should have no impact on this.

**Comment by ghorn** [comment]
How much effort do you estimate this is, and is there anything I can do to help as an outsider? This is making incremental builds extremely painful for my company.

**Comment by mgsloan** [comment]
@ghorn Good question.

1) Once https://github.com/haskell/cabal/pull/3066 is merged and a new version of Cabal is released, we can allow selectively building executables

2) Alternatively, we can revert the decision in https://github.com/commercialhaskell/stack/issues/1203 , and use `install` instead of `copy` + `register`.  This would be a behavior change, though, which will cause some projects which used to successfully build to fail, which is something we'd like to avoid.  Thoughts on this, @snoyberg / @borsboom ?  Do we want to preserve successful builds for funky projects even if it makes things less pleasant for normal projects?

So, in a way, we're in a corner where there's no good solution.  However, we could add a flag like `--cabal-setup-install` or something, to instead use `runhaskell Setup.hs install`, and also have this enable building executable components individually.  To save you from specifying this with every invocation of stack, https://github.com/commercialhaskell/stack/issues/846 can be implemented.

**Comment by snoyberg** [comment]
>  Alternatively, we can revert the decision in #1203 , and use install instead of copy + register. 

What leads you to believe that this will fix things? I've seen this bug appear with `runhaskell Setup.hs install`.

**Comment by mgsloan** [comment]
Good point, I haven't checked whether or not `install` also has this issue.  I believed it was this way because https://github.com/haskell/cabal/issues/2780 specifies `copy`.

**Comment by mgsloan** [comment]
This is particularly problematic for `stack ghci` - the library doesn't get registered if the copy step fails.  This means that if there's an executable with a compilation error, you can't load that executable into ghci :/

**Comment by mgsloan** [comment]
In theory, we could now lift this restriction when Cabal >= 1.24

**Comment by sjakobi** [comment]
> In theory, we could now lift this restriction when Cabal >= 1.24

Why not just error out when Cabal < 1.24 is installed and ask users to `stack setup --upgrade-cabal`?

**Comment by mgsloan** [comment]
@sjakobi Hmm, decent idea!  It seems like stack works fine with Cabal-1.24, but I don't think much scrutiny has been applied there.  `stack setup --upgrade-cabal` will already upgrade to that, though, so it'd better work.  I have noted that we should consider having an informational major version upper bound - https://github.com/commercialhaskell/stack/issues/2280#issuecomment-228252051

**Comment by sjakobi** [comment]
@mitchellwrosen has provided [an example project](https://github.com/mitchellwrosen/stack-issue) in #2366 for which the command `stack build stack-issue:lib` currently fails but should succeed once this issue is fixed.

**Comment by mgsloan** [comment]
I thought this made it into Cabal-1.24, but unfortunately, even though the change happened quite a while before that release, it is not included.

Unless we have some reason to believe a version of Cabal will have this fix in the near future, instead I am going to look into having stack always use a custom Setup.hs file which hacks around Cabal's deficiencies.

**Comment by mgsloan** [comment]
Once the next major version of Cabal is released, this can be resolved.  The docs on target syntax should be updated accordingly https://docs.haskellstack.org/en/stable/build_command/#target-syntax

**Comment by CMCDragonkai** [comment]
Is it possible to specify a package build target in the stack.yaml? I have package in `extra-deps`, that also produces an executable, the executable target has a dependency that is preventing me from resolving to a more recent LTS version.

**Comment by mgsloan** [comment]
@CMCDragonkai Not currently possible, but certainly worth considering, feel free to open an issue for it.

**Comment by stites** [comment]
@CMCDragonkai This is a problem for me as well -- I'm winding up having to configure my dependencies based on extra-dep executables.

**Comment by snoyberg** [comment]
PR #3237 (which is merged into master) allows building individual executables again. If you'd like to test out the functionality, give `stack upgrade --git` a shot. Since this issue is resolved on master, closing, thanks for the reports.

**Comment by saurabhnanda** [comment]
I ran into something similar, and want to confirm before posting to Stackoverflow. I have an executable which SHOULD NOT depend on, or build, the library. However, it seems that when I do `stack build webservice:exe:autogen` it ends up building the library target first. Is the behaviour related to this issue?

```
executable autogen
  main-is: GenerateModels.hs
  hs-source-dirs:
      scripts
  default-extensions: OverloadedStrings RecordWildCards DataKinds DeriveGeneric FlexibleInstances FlexibleContexts FunctionalDependencies GeneralizedNewtypeDeriving KindSignatures MultiParamTypeClasses TemplateHaskell TypeFamilies TypeSynonymInstances QuasiQuotes ScopedTypeVariables LambdaCase PolyKinds TypeApplications Arrows ConstraintKinds DuplicateRecordFields
  ghc-options: -threaded -rtsopts -with-rtsopts=-N -Wall -O0
  build-depends:
      base
  default-language: Haskell2010
```

**Comment by snoyberg** [comment]
That's impossible to avoid, Cabal (at least before 2.0) always builds the
library component

**Comment by saurabhnanda** [comment]
> That's impossible to avoid, Cabal (at least before 2.0) always builds the

Does this mean that Cabal 2.0+ will fix the unnecessary dependency on the library component? Or are you simply saying that you're sure of the pre-2.0 behaviour and unsure of the post-2.0 behaviour?

**Comment by snoyberg** [comment]
The latter, though I strongly believe Cabal 2.0 fixed it. We have plans
with the upcoming backpack support to modify Stack to take advantage of
this new functionality.

**Comment by saurabhnanda** [comment]
I did a `stack upgrade --git` and then tried re-building the executable via `stack build webservice:exe:autogen` and it still tried to build the library component first. So, I guess, Cabal-2.0+ also doesn't solve this.

**Comment by mgsloan** [comment]
@saurabhnanda Even if Cabal 2.0 has fixed the issue, there will need to be some changes to stack to fully fix the problem - https://github.com/commercialhaskell/stack/issues/3653 should definitely be looked into soon.

**Comment by aammfe** [comment]
I have two executables, 
1- a cross platform.
2- only available on linux and mac (bcs its uses unix package)

whenever I try to build it 1 at windows, eg: _stack build mypkg:exe:myExe1_  it tries to build 2 as well, which ultimately fails the build.

when I can expect this bug to resolve?

**Comment by mpilgrem** [comment]
@aammfe, the short answer to your question is, I don't know. I am not aware of anyone currently working on the matter. My own understanding of the underlying issue has not moved on since this comment: https://github.com/commercialhaskell/stack/issues/5647#issuecomment-1106796448. I will ask on the Haskell Community to see  if anyone knows if the upstream 'bug' in Cabal was ever resolved.
